bufferAppendError due to unknown issue with audio codec in Chrome browser

### What version of Hls.js are you using?

1.1.5

### What browser (including version) are you using?

Version 102.0.5005.61 (Official Build) (arm64)

### What OS (including version) are you using?

MacOS M1

### Test stream

https://hls-js.netlify.app/demo/?src=https%3A%2F%2Flive.hpdev.io%2Fvideos%2F12509_F6w16WnT7bjHeygc%2Fmaster.m3u8&demoConfig=eyJlbmFibGVTdHJlYW1pbmciOnRydWUsImF1dG9SZWNvdmVyRXJyb3IiOnRydWUsInN0b3BPblN0YWxsIjpmYWxzZSwiZHVtcGZNUDQiOmZhbHNlLCJsZXZlbENhcHBpbmciOi0xLCJsaW1pdE1ldHJpY3MiOi0xfQ==

### Configuration

```JavaScript
{
  "debug": true,
  "enableWorker": true,
  "lowLatencyMode": true,
  "backBufferLength": 90
}
```


### Additional player setup steps

_No response_

### Checklist

- [X] The issue observed is not already reported by searching on Github under https://github.com/video-dev/hls.js/issues
- [X] The issue occurs in the stable client (latest release) on https://hls-js.netlify.com/demo and not just on my page
- [X] The issue occurs in the latest client (main branch) on https://hls-js-dev.netlify.com/demo and not just on my page
- [X] The stream has correct Access-Control-Allow-Origin headers (CORS)
- [X] There are no network errors such as 404s in the browser console when trying to play the stream

### Steps to reproduce

1. use src link from example 
2. play stream with default HLS.js config in latest Chrome (in Safari works fine)


### Expected behaviour

Stream plays gracefully to the end like VLC does.

### What actually happened?

Playback hangs and freezes after a while (3-10 seconds, occasionally)

### Console output

```shell
buffer-controller.ts:407 [error] > [buffer-controller]: Failed 3 times to append segment in sourceBuffer
onError @ buffer-controller.ts:407
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1121
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
main.js:725 Error event: {type: 'mediaError', parent: 'main', details: 'bufferAppendError', err: DOMException: Failed to execute 'appendBuffer' on 'SourceBuffer': The HTMLMediaElement.error attrib…, fatal: true}
(anonymous) @ main.js:725
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
onError @ buffer-controller.ts:413
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1121
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
main.js:838 Fatal error : bufferAppendError
(anonymous) @ main.js:838
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
onError @ buffer-controller.ts:413
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1121
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
buffer-operation-queue.ts:62 [warn] > [buffer-operation-queue]: Unhandled exception executing the current operation
executeNext @ buffer-operation-queue.ts:62
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1145
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
buffer-controller.ts:384 [error] > [buffer-controller]: Error encountered while trying to append to the audio SourceBuffer DOMException: Failed to execute 'appendBuffer' on 'SourceBuffer': The HTMLMediaElement.error attribute is not null.
    at BufferController.appendExecutor (https://hls-js.netlify.app/dist/hls.js:4968:8)
    at Object.execute (https://hls-js.netlify.app/dist/hls.js:4507:16)
    at BufferOperationQueue.executeNext (https://hls-js.netlify.app/dist/hls.js:5113:19)
    at BufferOperationQueue.append (https://hls-js.netlify.app/dist/hls.js:5076:12)
    at BufferController.onBufferAppending (https://hls-js.netlify.app/dist/hls.js:4573:20)
    at EventEmitter.emit (https://hls-js.netlify.app/dist/hls.js:310:33)
    at Hls.emit (https://hls-js.netlify.app/dist/hls.js:17266:26)
    at Hls.trigger (https://hls-js.netlify.app/dist/hls.js:17271:19)
    at StreamController.bufferFragmentData (https://hls-js.netlify.app/dist/hls.js:3554:14)
    at StreamController._handleTransmuxComplete (https://hls-js.netlify.app/dist/hls.js:10360:12)
onError @ buffer-controller.ts:384
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1145
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
buffer-controller.ts:407 [error] > [buffer-controller]: Failed 3 times to append segment in sourceBuffer
onError @ buffer-controller.ts:407
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1145
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
main.js:725 Error event: {type: 'mediaError', parent: 'main', details: 'bufferAppendError', err: DOMException: Failed to execute 'appendBuffer' on 'SourceBuffer': The HTMLMediaElement.error attrib…, fatal: true}
(anonymous) @ main.js:725
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
onError @ buffer-controller.ts:413
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1145
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
main.js:838 Fatal error : bufferAppendError
(anonymous) @ main.js:838
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
onError @ buffer-controller.ts:413
executeNext @ buffer-operation-queue.ts:65
append @ buffer-operation-queue.ts:25
onBufferAppending @ buffer-controller.ts:416
emit @ index.js:203
emit @ hls.ts:250
trigger @ hls.ts:258
bufferFragmentData @ base-stream-controller.ts:719
_handleTransmuxComplete @ stream-controller.ts:1145
handleTransmuxComplete @ transmuxer-interface.ts:315
onWorkerMessage @ transmuxer-interface.ts:281
base-stream-controller.ts:1381 [log] > [stream-controller]: PARSING->PARSED
base-stream-controller.ts:499 [log] > [stream-controller]: Buffered main sn: 5 of level 4 [2.020,12.000]
base-stream-controller.ts:1381 [log] > [stream-controller]: PARSED->IDLE
```


### Chrome media internals output

_No response_

--- a/src/demux/adts.ts
+++ b/src/demux/adts.ts
@@ -23,7 +23,6 @@ type AudioConfig = {
 type FrameHeader = {
   headerLength: number;
   frameLength: number;
-  stamp: number;
 };
 
 export function getAudioConfig(
@@ -255,21 +254,17 @@ export function getFrameDuration(samplerate: number): number {
 
 export function parseFrameHeader(
   data: Uint8Array,
-  offset: number,
-  pts: number,
-  frameIndex: number,
-  frameDuration: number
+  offset: number
 ): FrameHeader | void {
   // The protection skip bit tells us if we have 2 bytes of CRC data at the end of the ADTS header
   const headerLength = getHeaderLength(data, offset);
-  // retrieve frame size
-  let frameLength = getFullFrameLength(data, offset);
-  frameLength -= headerLength;
-
-  if (frameLength > 0) {
-    const stamp = pts + frameIndex * frameDuration;
-    // logger.log(`AAC frame, offset/length/total/pts:${offset+headerLength}/${frameLength}/${data.byteLength}/${(stamp/90).toFixed(0)}`);
-    return { headerLength, frameLength, stamp };
+  if (offset + headerLength <= data.length) {
+    // retrieve frame size
+    const frameLength = getFullFrameLength(data, offset) - headerLength;
+    if (frameLength > 0) {
+      // logger.log(`AAC frame, offset/length/total/pts:${offset+headerLength}/${frameLength}/${data.byteLength}`);
+      return { headerLength, frameLength };
+    }
   }
 }
 
@@ -279,15 +274,16 @@ export function appendFrame(
   offset: number,
   pts: number,
   frameIndex: number
-): AudioFrame | void {
+): AudioFrame {
   const frameDuration = getFrameDuration(track.samplerate as number);
-  const header = parseFrameHeader(data, offset, pts, frameIndex, frameDuration);
+  const stamp = pts + frameIndex * frameDuration;
+  const header = parseFrameHeader(data, offset);
+  let unit: Uint8Array;
   if (header) {
-    const { frameLength, headerLength, stamp } = header;
+    const { frameLength, headerLength } = header;
     const length = headerLength + frameLength;
     const missing = Math.max(0, offset + length - data.length);
     // logger.log(`AAC frame ${frameIndex}, pts:${stamp} length@offset/total: ${frameLength}@${offset+headerLength}/${data.byteLength} missing: ${missing}`);
-    let unit: Uint8Array;
     if (missing) {
       unit = new Uint8Array(length - headerLength);
       unit.set(data.subarray(offset + headerLength, data.length), 0);
@@ -305,4 +301,13 @@ export function appendFrame(
 
     return { sample, length, missing };
   }
+  // overflow incomplete header
+  const length = data.length - offset;
+  unit = new Uint8Array(length);
+  unit.set(data.subarray(offset, data.length), 0);
+  const sample: AudioSample = {
+    unit,
+    pts: stamp,
+  };
+  return { sample, length, missing: -1 };
 }

--- a/src/demux/tsdemuxer.ts
+++ b/src/demux/tsdemuxer.ts
@@ -837,20 +837,26 @@ class TSDemuxer implements Demuxer {
   private parseAACPES(track: DemuxedAudioTrack, pes: PES) {
     let startOffset = 0;
     const aacOverFlow = this.aacOverFlow;
-    const data = pes.data;
+    let data = pes.data;
     if (aacOverFlow) {
       this.aacOverFlow = null;
+      const frameMissingBytes = aacOverFlow.missing;
       const sampleLength = aacOverFlow.sample.unit.byteLength;
-      const frameMissingBytes = Math.min(aacOverFlow.missing, sampleLength);
-      const frameOverflowBytes = sampleLength - frameMissingBytes;
-      aacOverFlow.sample.unit.set(
-        data.subarray(0, frameMissingBytes),
-        frameOverflowBytes
-      );
-      track.samples.push(aacOverFlow.sample);
-
-      // logger.log(`AAC: append overflowing ${frameOverflowBytes} bytes to beginning of new PES`);
-      startOffset = aacOverFlow.missing;
+      // logger.log(`AAC: append overflowing ${sampleLength} bytes to beginning of new PES`);
+      if (frameMissingBytes === -1) {
+        const tmp = new Uint8Array(sampleLength + data.byteLength);
+        tmp.set(aacOverFlow.sample.unit, 0);
+        tmp.set(data, sampleLength);
+        data = tmp;
+      } else {
+        const frameOverflowBytes = sampleLength - frameMissingBytes;
+        aacOverFlow.sample.unit.set(
+          data.subarray(0, frameMissingBytes),
+          frameOverflowBytes
+        );
+        track.samples.push(aacOverFlow.sample);
+        startOffset = aacOverFlow.missing;
+      }
     }
     // look for ADTS header (0xFFFx)
     let offset: number;
@@ -906,26 +912,20 @@ class TSDemuxer implements Demuxer {
 
     // scan for aac samples
     let frameIndex = 0;
+    let frame;
     while (offset < len) {
-      if (ADTS.isHeader(data, offset)) {
-        if (offset + 5 < len) {
-          const frame = ADTS.appendFrame(track, data, offset, pts, frameIndex);
-          if (frame) {
-            if (frame.missing) {
-              this.aacOverFlow = frame;
-            } else {
-              offset += frame.length;
-              frameIndex++;
-              continue;
-            }
+      frame = ADTS.appendFrame(track, data, offset, pts, frameIndex);
+      offset += frame.length;
+      if (!frame.missing) {
+        frameIndex++;
+        for (; offset < len - 1; offset++) {
+          if (ADTS.isHeader(data, offset)) {
+            break;
           }
         }
-        // We are at an ADTS header, but do not have enough data for a frame
-        // Remaining data will be added to aacOverFlow
-        break;
       } else {
-        // nothing found, keep looking
-        offset++;
+        this.aacOverFlow = frame;
+        break;
       }
     }
   }

--- a/tests/unit/demuxer/adts.js
+++ b/tests/unit/demuxer/adts.js
@@ -374,10 +374,9 @@ describe('parseFrameHeader', function () {
     data[0] = 0xff;
     data[1] = 0xf0; // protection_absent = 0
     data[4] = 0x02; // frame_length is 16
-    expect(parseFrameHeader(data, 0, 0, 0, 0)).to.deep.equal({
+    expect(parseFrameHeader(data, 0)).to.deep.equal({
       headerLength: 9,
       frameLength: 7,
-      stamp: 0,
     });
   });
 
@@ -386,12 +385,12 @@ describe('parseFrameHeader', function () {
     data[0] = 0xff;
     data[1] = 0xf0; // protection_absent = 0
     data[4] = 0x00; // frame_length is 0
-    expect(parseFrameHeader(data, 0, 0, 0, 0)).to.be.undefined;
+    expect(parseFrameHeader(data, 0)).to.be.undefined;
   });
 });
 
 describe('appendFrame', function () {
-  it('should append the found sample to track and return some useful information', function () {
+  it('should append the found sample to track and return frame information', function () {
     const track = {
       samplerate: 64000,
       samples: [],
@@ -414,7 +413,7 @@ describe('appendFrame', function () {
     expect(track.samples.length).to.equal(1);
   });
 
-  it('should not append sample when incomplete (aac overflow or progressive streaming)', function () {
+  it('should return an incomplete frame without appending when data is incomplete (aac overflow or progressive streaming)', function () {
     const track = {
       samplerate: 64000,
       samples: [],
@@ -436,6 +435,31 @@ describe('appendFrame', function () {
       },
       length: 24,
       missing: 4,
+    });
+    expect(track.samples.length).to.equal(0);
+  });
+
+  it('should return an incomplete frame without appending when header is incomplete (aac overflow or progressive streaming)', function () {
+    const track = {
+      samplerate: 64000,
+      samples: [],
+      len: 0,
+    };
+    const data = new Uint8Array(new ArrayBuffer(2));
+    data[0] = 0xff;
+    data[1] = 0xf0; // protection_absent = 0
+
+    const frame = appendFrame(track, data, 0, 0, 0);
+    const unit = new Uint8Array(2);
+    unit.set(data.subarray(0, 2), 0);
+
+    expect(frame, JSON.stringify(frame)).to.deep.equal({
+      sample: {
+        unit,
+        pts: 0,
+      },
+      length: 2,
+      missing: -1,
     });
     expect(track.samples.length).to.equal(0);
   });

